Light client tutorial using `bdk-cli`

rajarshimaitra opened the pull request:
This is a draft tutorial I am working on for a demo of `bdk-cli` with `compact_filters` feature. (included in https://github.com/bitcoindevkit/bdk-cli/pull/26)

This allows wallet devs to easily spawn their Neurtino like backend using BDK.

The demo only covers basic wallet functionalities. 

The demo needs `bdk-cli` compiled on the above PR branch.

Suggestions, improvements, and additions are welcome.

### Note to Reviewers.
Until https://github.com/bitcoindevkit/bdk-cli/pull/26 gets merged, you have to build `bdk-cli` on top of that PR branch to get `compact_filters` support.

rajarshimaitra committed:
Add author details

rajarshimaitra committed:
Add tutorial text

i5hi commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+- Install `bdk-cli` with `compact_filters` feature
+
+  ```shell
+  $ cargo install bdk-cli --features compact_filters
````
Package `bdk-cli v0.2.0` does not have the feature `compact_filters`

rajarshimaitra replied:
Yes, because https://github.com/bitcoindevkit/bdk-cli/pull/26 hasn't been merged yet. You can build `bdk-cli` on top of that PR branch to get `compact_filters` support.

i5hi replied:
oops, sorry!

i5hi replied:
got the correct version from your pr branch now :P

notmandatory commented on content/authors/rajarshi-maitra/_index.md:
```diff
@@ -0,0 +1,8 @@
+name: Rajarshi Maitra
+photo: 
+twitter: rajarshimaitra
+mastodon: https://bitcoinhackers.org/@rajarshimaitra
```
Should be:
```suggestion
mastodon: https://bitcoinhackers.org/@raj
```

rajarshimaitra replied:
Huh, seems like I forgot my own handle. Thanks.. :)

notmandatory commented on content/blog/2021/compact_filters_demo.md:
```diff
@@ -0,0 +1,367 @@
+## Introduction
+
+#### Compact Filters:
+Compact filters are the latest specification of Bitcoin SPV node implementation as per [BIP157](https://github.com/bitcoin/bips/blob/master/bip-0157.mediawiki) and [BIP158](https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki). These filters allow creation of light client nodes (aka Neutrinos). Such light clients have been envisioned in the white paper also by Satoshi himself, but due t ...
```
I think it'd be clearer instead of saying "(aka Neutrinos)" .. and also a little rewording,  something like:
```suggestion
Compact filters are the latest specification of Bitcoin SPV node implementation as per [BIP157](https://github.com/bitcoin/bips/blob/master/bip-0157.mediawiki) and [BIP158](https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki). The [Neutrino](https://github.com/lightninglabs/neutrino) project pioneered the use of compact filter based light client nodes for use with Lightning Network wallets. Such light clients were envisioned by Satoshi himself in  his original white paper, but due to lack of robust privacy and trust guarantees using conventional [bloomfilters](https://github.com/bitcoin/bips/blob/master/bip-0037.mediawiki), these type of nodes never got popular. 
```

rajarshimaitra replied:
Seems good..

notmandatory commented on content/blog/2021/compact_filters_demo.md:
```diff
@@ -0,0 +1,367 @@
+description: "Tutorial showing usage of compact filters (BIP157) using bdk-cli command line tools"
+authors:
+    - Rajarshi Maitra
+date: "2021-05-28"
```
Best to update this to:
```suggestion
date: "2021-06-17"
```

rajarshimaitra replied:
Done.

notmandatory commented on content/blog/2021/compact_filters_demo.md:
```diff
@@ -0,0 +1,367 @@
+Bitcoin core supports `BIP157` from `v0.21.0`.  
+
+#### BDK and Compact filters
+BDK is a bitcoin wallet development library that can be used to create bitcoin wallets with ith custom `Database` and `Blockchain` backend. BDK is a [descriptor](https://bitcoindevkit.org/descriptors/) based wallet, i.e. the wallet keychain is described by a set of descriptors. 
```
```suggestion
BDK is a bitcoin wallet development library that can be used to create bitcoin wallets with custom `Database` and `Blockchain` backends. BDK is a [descriptor](https://bitcoindevkit.org/descriptors/) based wallet, i.e. the wallet keychain is described by a set of descriptors. 
```

rajarshimaitra replied:
Done.

notmandatory commented on content/blog/2021/compact_filters_demo.md:
```diff
@@ -0,0 +1,367 @@
+#### Compact Filters:
+Compact filters are the latest specification of Bitcoin SPV node implementation as per [BIP157](https://github.com/bitcoin/bips/blob/master/bip-0157.mediawiki) and [BIP158](https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki). These filters allow creation of light client nodes (aka Neutrinos). Such light clients have been envisioned in the white paper also by Satoshi himself, but due t ...
+
+Enters [BIP157](https://github.com/bitcoin/bips/blob/master/bip-0157.mediawiki), which described a new type of filters for Bitcoin Blockchain data, known as `compact_filters`. Using compact filters, a light-node can talk to one or more full nodes, and fetch relevant information on the blockchain. With much more robust privacy and security guarantees than previously possible. These are **Neutrinos ...
```
I also recommend rewording here since I don't think CBF clients are necessarily referred to as "Neutrinos" that project was just the first implementation of the BIPs.

```suggestion
Enters [BIP157](https://github.com/bitcoin/bips/blob/master/bip-0157.mediawiki), which described a new type of filters for Bitcoin Blockchain data, known as `compact_filters`. Using compact filters, a light-node can talk to one or more full nodes, and fetch relevant information on the blockchain. With much more robust privacy and security guarantees than previously possible. Compact filter based nodes are best suitable to be used with mobile wallets, to create more trustless mobile applications on Bitcoin. Any wallet application that needs to have an "eye on the blockchain" has an use for such light clients. 
```

rajarshimaitra replied:
Ya, I thought they were called Neutrino in general. Updated.

notmandatory commented on content/blog/2021/compact_filters_demo.md:
```diff
@@ -0,0 +1,367 @@
+#### BDK and Compact filters
+BDK is a bitcoin wallet development library that can be used to create bitcoin wallets with ith custom `Database` and `Blockchain` backend. BDK is a [descriptor](https://bitcoindevkit.org/descriptors/) based wallet, i.e. the wallet keychain is described by a set of descriptors. 
+
+Using BDK one can instantiate wallets of various kinds as per requirement. BDK abstracts away all the heavy lifting works, and allow wallet devs to concentrate on logic that they care about, i.e. writing wallet codes. For more detailed documentation on BDK capabilities check these [blog](https://bitcoindevkit.org/blog/2020/12/hello-world/), [bog](https://bitcoindevkit.org/blog/2020/11/descriptors ...
```
If you leave out the version then the latest will be used:
```suggestion
Using BDK one can instantiate wallets of various kinds as per requirement. BDK abstracts away all the heavy lifting works, and allow wallet devs to concentrate on logic that they care about, i.e. writing wallet codes. For more detailed documentation on BDK capabilities check these [blog](https://bitcoindevkit.org/blog/2020/12/hello-world/), [bog](https://bitcoindevkit.org/blog/2020/11/descriptors-in-the-wild/) and [docs](https://docs.rs/bdk/).
```

rajarshimaitra replied:
Updated.

notmandatory commented on content/blog/2021/compact_filters_demo.md:
```diff
@@ -0,0 +1,367 @@
+
+BDK comes with default implementations of all them that developers can start with. Developers can also create there own custom implementations and plug it into BDK (thanks to rust magic of `Traits`).
+
+BDK implements BIP158 communication protocol, that wallet devs can use to create their own Neutrino node that can request and receive filter data from one or many full nodes. This capability can be unlocked by using BDK with `compact_filters` feature. Once enabled, BDK will be able to create wallets with the `compact_filters` type `Blockchain` backend. (The default `Blockchain` is a electrum serv ...
```
small nits:
```suggestion
BDK implements the BIP158 communication protocol, that wallet devs can use to create their own Neutrino node that can request and receive filter data from one or many full nodes. This capability can be unlocked by using BDK with the `compact_filters` feature. Once enabled, BDK will be able to create wallets with the `compact_filters` type `Blockchain` backend. (The default `Blockchain` is a electrum server) 
```

rajarshimaitra replied:
Done.

notmandatory commented on content/blog/2021/compact_filters_demo.md:
```diff
@@ -0,0 +1,367 @@
+BDK implements BIP158 communication protocol, that wallet devs can use to create their own Neutrino node that can request and receive filter data from one or many full nodes. This capability can be unlocked by using BDK with `compact_filters` feature. Once enabled, BDK will be able to create wallets with the `compact_filters` type `Blockchain` backend. (The default `Blockchain` is a electrum serv ...
+
+#### bdk-cli
+`bdk-cli` is lightweight [REPL](https://codewith.mu/en/tutorials/1.0/repl) wrapper over the BDK library to facilitate quick and easy demonstration of BDK capabilities in command-line. Wallet devs can use this tool to quickly try out different possibilities with BDK.
```
nit:
```suggestion
`bdk-cli` is a lightweight [REPL](https://codewith.mu/en/tutorials/1.0/repl) wrapper over the BDK library to facilitate quick and easy demonstration of BDK capabilities in command-line. Wallet devs can use this tool to quickly try out different possibilities with BDK.
```

rajarshimaitra replied:
Done.

notmandatory commented on content/blog/2021/compact_filters_demo.md:
```diff
@@ -0,0 +1,367 @@
+
+You also need to ensure proper configuration settings for signalling `compact_filters` support.  
+
+For ease of testing, BDK project hosts docker images that can be used to spawn Bitcoin Core with all the relevant configurations.
```
nit:
```suggestion
For ease of testing, the BDK project hosts docker images that can be used to spawn Bitcoin Core with all the relevant configurations.
```

rajarshimaitra replied:
Done.

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+  This should start logging the `debug.log` in the terminal. You can keep this terminal alive to see communication events with BDK and the node.
+
+  ```shell
+  $ docker run -p 127.0.0.1:18443-18444:18443-18444/tcp --name bdk-box bitcoindevkit/bitcoind
````
This will put the container in the background and automatically clean it up when killed:
```suggestion
  $ docker run --detach --rm -p 127.0.0.1:18443-18444:18443-18444/tcp --name bdk-box bitcoindevkit/bitcoind
  
  # view logs
  $ docker logs bdk-box
```

rajarshimaitra replied:
Done.

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+  Assuming you have `bitcoin-cli` already installed, 
+  In another terminal try connecting to the node with `bitcoin-cli`
+  ```shell
+  $ bitcoin-cli getnetworkinfo
````
I think it'd be easier to not need bitcoin-cli installed this will work:
```suggestion
  $ docker exec -it bdk-box /root/bitcoin-cli -regtest getnetworkinfo
```

rajarshimaitra replied:
Makes sense..

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+  ```
+- Check node is reachable
+
+  Assuming you have `bitcoin-cli` already installed, 
````
If you are OK using the `bdk-box` for the `bitcoin-cli` you can remove this line.

rajarshimaitra replied:
Removed. No need for using `bitcoin-cli` separately, when it already comes with the docker.

notmandatory commented on content/blog/2021/compact_filters_demo.md:
```diff
@@ -0,0 +1,367 @@
+  $ bitcoin-cli getnetworkinfo
+  {
+    "version": 210000,
+    "subversion": "/Satoshi:0.21.0/",
```
I recently updated the docker images, so version changed:
```suggestion
    "subversion": "/Satoshi:0.21.1/",
```

rajarshimaitra replied:
Done

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+- Install `bdk-cli` with `compact_filters` feature
+
+  ```shell
+  $ cargo install bdk-cli --features compact_filters
````
If we use this command we won't have to wait for the new `3.0` version before publishing this blog:
```suggestion
  $ cargo install --git https://github.com/bitcoindevkit/bdk-cli.git bdk-cli --features compact_filters
```

rajarshimaitra replied:
Ok seems good. Updated.

notmandatory commented on content/blog/2021/compact_filters_demo.md:
```diff
@@ -0,0 +1,367 @@
+
+  We will use the following paths for our `receive` and `change` descriptors 
+
+	- `receive` path: `m/84h/0h/0h/0` 
```
Technically since we're using regtest/testnet the coin type for BIP84 should be 1h (as per [BIP44](https://en.bitcoin.it/wiki/BIP_0044#Registered_coin_types)):
```suggestion
	- `receive` path: `m/84h/1h/0h/0` 
```

rajarshimaitra replied:
Ah thats right..

notmandatory commented on content/blog/2021/compact_filters_demo.md:
```diff
@@ -0,0 +1,367 @@
+  We will use the following paths for our `receive` and `change` descriptors 
+
+	- `receive` path: `m/84h/0h/0h/0` 
+	- `change` path: `m/84h/0h/0h/1`, 
```
```suggestion
	- `change` path: `m/84h/1h/0h/1`, 
```

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+  When asking for a new address BDK will derive one from the `receive` descriptor, and while constructing transaction, BDK will use the `change` descriptor to derive change address.
+
+	```shell
+  $ BDK_recv_desc="wpkh($(bdk-cli key derive --path m/84h/0h/0h/0 --xprv $BDK_xprv | jq -r '.xprv'))"
````
```suggestion
  $ BDK_recv_desc="wpkh($(bdk-cli key derive --path m/84h/1h/0h/0 --xprv $BDK_xprv | jq -r '.xprv'))"
```

rajarshimaitra replied:
Done..

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+	```shell
+  $ BDK_recv_desc="wpkh($(bdk-cli key derive --path m/84h/0h/0h/0 --xprv $BDK_xprv | jq -r '.xprv'))"
+  $ echo $BDK_recv_desc
+  wpkh([ff09c7c9/84'/0'/0'/0]tprv8hkdEGgwLLnqsdfkJFidpTj5d6z5qFdP6Qwzsviea3HrS9C2mXXaDivPKCCgcaWvnGNX9eciLUQs91PWYXJqrChfnAagViCgG6L5phaNyWr/*) 
````
```suggestion
  wpkh([ff09c7c9/84'/1'/0'/0]tprv8hkdEGgwLLnqsdfkJFidpTj5d6z5qFdP6Qwzsviea3HrS9C2mXXaDivPKCCgcaWvnGNX9eciLUQs91PWYXJqrChfnAagViCgG6L5phaNyWr/*) 
```

rajarshimaitra replied:
done

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+  wpkh([ff09c7c9/84'/0'/0'/0]tprv8hkdEGgwLLnqsdfkJFidpTj5d6z5qFdP6Qwzsviea3HrS9C2mXXaDivPKCCgcaWvnGNX9eciLUQs91PWYXJqrChfnAagViCgG6L5phaNyWr/*) 
+  ```
+  ```shell
+  $ BDK_chng_desc="wpkh($(bdk-cli key derive --path m/84h/0h/0h/1 --xprv $BDK_xprv | jq -r '.xprv'))"
````
```suggestion
  $ BDK_chng_desc="wpkh($(bdk-cli key derive --path m/84h/1h/0h/1 --xprv $BDK_xprv | jq -r '.xprv'))"
```

rajarshimaitra replied:
Done..

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+  ```shell
+  $ BDK_chng_desc="wpkh($(bdk-cli key derive --path m/84h/0h/0h/1 --xprv $BDK_xprv | jq -r '.xprv'))"
+  $ echo $BDK_chng_desc 
+  wpkh([ff09c7c9/84'/0'/0'/1]tprv8hkdEGgwLLnqtbYkGG7fSy7v43RF2SQGGjNuZtmBzEHh7H8xgpXBETQAbVPqi8rkvLNFKLYY4rDzXA4fn5Ha1yuazZqhQPe3uNKmFS7648s/*)
````
```suggestion
  wpkh([ff09c7c9/84'/1'/0'/1]tprv8hkdEGgwLLnqtbYkGG7fSy7v43RF2SQGGjNuZtmBzEHh7H8xgpXBETQAbVPqi8rkvLNFKLYY4rDzXA4fn5Ha1yuazZqhQPe3uNKmFS7648s/*)
```

rajarshimaitra replied:
done

notmandatory commented on content/blog/2021/compact_filters_demo.md:
```diff
@@ -0,0 +1,367 @@
+  
+   
+ 
+### Bitcoin Core Wallet Generation
```
The "Bitcoin Core Wallet Generation" section needs to be moved  above the "Create and Sync a wallet" section because compact block filters won't be served and the sync will fail unless some blocks have been generated by bitcoind first.

rajarshimaitra replied:
Yes that seems better..

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+
+- Create a wallet and generate 101 blocks. 
+  ```shell
+  $ bitcoin-cli createwallet test
````
```suggestion
  $ docker exec -it bdk-box /root/bitcoin-cli -regtest createwallet test
```

rajarshimaitra replied:
done

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+  ```
+
+  ```shell
+  $ bitcoin-cli getnewaddress
````
```suggestion
  $ docker exec -it bdk-box /root/bitcoin-cli -regtest getnewaddress
```

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+  bcrt1qatd7yq0jukwusuaufltlejmeydpvnpv43r5gc2
+  ```
+  ```shell
+  $ bitcoin-cli generatetoaddress 101 bcrt1qatd7yq0jukwusuaufltlejmeydpvnpv43r5gc2
````
```suggestion
  $ docker exec -it bdk-box /root/bitcoin-cli -regtest generatetoaddress 101 bcrt1qatd7yq0jukwusuaufltlejmeydpvnpv43r5gc2
```

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+  $ bitcoin-cli getbalance
+  50.00000000
+  ```
+- Now the core wallet is loaded and funded so we can use it to make transactions.
````
```suggestion
- Now the core wallet has generated new blocks and is funded with test bitcoin so we can sync our bdk wallet and use it to make transactions.
```

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+  ]
+  ```
+  ```shell
+  $ bitcoin-cli getbalance
````
```suggestion
  $ docker exec -it bdk-box /root/bitcoin-cli -regtest getbalance
```

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+
+- Transfer funds to the previous address and generate a block, using `bitcoin-cli`
+  ```shell
+  $ bitcoin-cli sendtoaddress bcrt1qx2479wywulf50pqx5uy64zhxq9f3tuvlh8u0s9 5
````
```suggestion
  $ docker exec -it bdk-box /root/bitcoin-cli -regtest sendtoaddress bcrt1qx2479wywulf50pqx5uy64zhxq9f3tuvlh8u0s9 5
```

notmandatory commented on content/blog/2021/compact_filters_demo.md:
```diff
@@ -0,0 +1,367 @@
+  $ bitcoin-cli sendtoaddress bcrt1qx2479wywulf50pqx5uy64zhxq9f3tuvlh8u0s9 5
+
+
+  $ bitcoin-cli generatetoaddress 1 bcrt1qw3ht9xtc9pgyvmqay0ap9fw8mxd27az8el0uz3
```
```suggestion
  $ docker exec -it bdk-box /root/bitcoin-cli -regtest generatetoaddress 1 bcrt1qw3ht9xtc9pgyvmqay0ap9fw8mxd27az8el0uz3
```

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+
+- Check transaction in mempool
+  ```shell
+  $ bitcoin-cli gettransaction c343f5b25372e285308eba912d1fe8fade9f64afde6d95306e2248e52e0852252
````
```suggestion
  $ docker exec -it bdk-box /root/bitcoin-cli -regtest gettransaction c343f5b25372e285308eba912d1fe8fade9f64afde6d95306e2248e52e0852252
```

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+
+- Generate 1 block to confirm the transaction
+  ```shell
+  $ bitcoin-cli generatetoaddress 1 bcrt1qatd7yq0jukwusuaufltlejmeydpvnpv43r5gc2
````
```suggestion
  $ docker exec -it bdk-box /root/bitcoin-cli -regtest generatetoaddress 1 bcrt1qatd7yq0jukwusuaufltlejmeydpvnpv43r5gc2
```

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+
+- fetch a new `core` address
+  ```shell
+  $ core_addrs=$(bitcoin-cli getnewaddress)
````
```suggestion
  $ core_addrs=$(docker exec -it bdk-box /root/bitcoin-cli -regtest getnewaddress | tr -d '\r')
```

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+
+- Sync the `bdk-test` wallet and ask for available balance.
+  ```shell
+	$ bdk-cli --network regtest wallet --node 127.0.0.1:18444 --wallet bdk-test --descriptor $desc sync
````
```suggestion
	$ bdk-cli --network regtest wallet --node "127.0.0.1:18444" --wallet bdk-test -d $BDK_recv_desc -c $BDK_chng_desc sync
```

rajarshimaitra replied:
Eh missed that..

notmandatory commented on content/blog/2021/compact_filters_demo.md:
```diff
@@ -0,0 +1,367 @@
+	$ bdk-cli --network regtest wallet --node 127.0.0.1:18444 --wallet bdk-test --descriptor $desc sync
+	{}
+	
+	$ bdk-cli --network regtest wallet --node 127.0.0.1:18444 --wallet bdk-test --descriptor $desc get_balance
```
```suggestion
	$ bdk-cli --network regtest wallet --node "127.0.0.1:18444" --wallet bdk-test -d $BDK_recv_desc -c $BDK_chng_desc get_balance
```

notmandatory reviewed (changes requested):
Hi sorry for the delay in getting this reviewed. I ran through all the steps and found a few things that should be fixed. For the `bitcoin-cli` commands I switched them to use the internal docker command so we don't need to worry about bitcoind rpc permissions.

notmandatory commented on content/blog/2021/compact_filters_demo.md:
````diff
@@ -0,0 +1,367 @@
+  ```
+
+	Voila! We have the updated balance. 
+
````
I would also add a step to kill the container, so it will get cleaned up in case they want to run the tutorial again, something like: 

You may now shutdown the regtest docker container, this will also clean up any created data.

```
  $ docker kill bdk-box
```

rajarshimaitra replied:
Ya makes sense..

rajarshimaitra committed:
Various nits and fixes from @notmandatory

rajarshimaitra commented:
Thanks @notmandatory for taking the time for such a thorough review.

Addressed all the comments. Hope I didn't miss any.

notmandatory reviewed (approved):
Looks good, I know these take a lot of work so thanks for putting it all together.

notmandatory commented:
Looks good, I know these take a lot of work so thanks for putting it all together.  Please move it back to "Ready for review" when you're ready to have it merged.

rajarshimaitra committed:
update various nits

Updated with better structuring of various sentences and segments.

Removed mentions of Neutrino apart from metadata and credit.

rajarshimaitra commented:
Thanks @notmandatory , I think I will always find nits to update every time I look. 

Made a bunch of modification. Removed mentions of Neutrino in general and called it `comapct_filters`. Apart from metadata, as Neutrino is a popular term, so it might help in google indexing.

So far this seems final from my side. 

Opening for the merge.

notmandatory reviewed (approved):
ACK, I also ran through the examples again and confirmed they work.

notmandatory commented:
I had to also do a re-base to resolve unrelated merge conflicts.
